modules/src/ics04_channel/context.rs

```diff
@@ -20,51 +21,68 @@ pub trait ChannelReader {
+    /// Returns the ClientState for the given identifier `client_id`. Necessary dependency towards
+    /// proof verification.
+    fn client_state(&self, client_id: &ClientId) -> Option<AnyClientState>;
 
-    fn channel_client_consensus_state(
+    fn client_consensus_state(
         &self,
-        port_channel_id: &(PortId, ChannelId),
+        client_id: &ClientId,
         height: Height,
     ) -> Option<AnyConsensusState>;
 
     fn authenticated_capability(&self, port_id: &PortId) -> Result<Capability, Error>;
+
+    /// Returns a counter on the number of channel ids have been created thus far.
+    /// The value of this counter should increase only via method
+    /// `ChannelKeeper::increase_channel_counter`.
+    fn channel_counter(&self) -> u64;
 }
 
 /// A context supplying all the necessary write-only dependencies (i.e., storage writing facility)
 /// for processing any `ChannelMsg`.
 pub trait ChannelKeeper {
-    fn next_channel_id(&mut self) -> ChannelId;
-
     fn store_channel_result(&mut self, result: ChannelResult) -> Result<(), Error> {
-        if result.channel_id.is_none() {
-            // If this is the first time the handler processed this channel
-            let channel_id = self.next_channel_id();
-
-            self.store_channel(
-                &(result.port_id.clone(), channel_id.clone()),
-                &result.channel_end,
-            )?;
-
-            // associate also the channel end to its connection
-            self.store_connection_channels(
-                &result.channel_end.connection_hops()[0].clone(),
-                &(result.port_id.clone(), channel_id.clone()),
-            )?;
-
-            // initialize send sequence number
-            self.store_next_sequence_send(&(result.port_id.clone(), channel_id.clone()), 1)?;
-            // initialize recv sequence number
-            self.store_next_sequence_recv(&(result.port_id.clone(), channel_id.clone()), 1)?;
-            // initialize ack sequence number
-            self.store_next_sequence_ack(&(result.port_id, channel_id), 1)?;
-        } else {
-            //the handler processed this channel for channel open init
-            self.store_channel(
-                &(result.port_id.clone(), result.channel_id.clone().unwrap()),
-                &result.channel_end,
-            )?;
+        match result.channel_end.state() {
+            // This is the first time the handler processed this channel, with `MsgChannelOpenInit`.
+            State::Init => {
+                self.store_channel(
+                    &(result.port_id.clone(), result.channel_id.clone()),
+                    &result.channel_end,
+                )?;
+
+                // Associate also the channel end to its connection.
+                self.store_connection_channels(
+                    &result.channel_end.connection_hops()[0].clone(),
+                    &(result.port_id.clone(), result.channel_id.clone()),
+                )?;
+
```

Comment:
I liked it better before :) I don't think we should read too deeply into the channel state for everything.  We should not be trying to validate what `process` did here (*). Same for connection and I provided similar comments there. 
What happens for `State::TryOpen`? don't we have a similar case as for Init?
I think we should just store the channel and store it under connection regardless of state. Increase the channel counter if previous_channel_id is None. Store the sequences only for `Init` state. 

(*) if we do, then define different result types for each message.

Reply:
Refactor with a simpler approach.